sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala

```diff
@@ -134,8 +142,26 @@ object FileFormatWriter extends Logging {
       // prepares the job, any exception thrown from here shouldn't cause abortJob() to be called.
       committer.setupJob(job)
 
+      val bucketIdExpression = bucketSpec.map { spec =>
+        // Use `HashPartitioning.partitionIdExpression` as our bucket id expression, so that we can
+        // guarantee the data distribution is same between shuffle and bucketed data source, which
+        // enables us to only shuffle one side when join a bucketed table and a normal one.
+        HashPartitioning(bucketColumns, spec.numBuckets).partitionIdExpression
+      }
+      // We should first sort by partition columns, then bucket id, and finally sorting columns.
+      val requiredOrdering = (partitionColumns ++ bucketIdExpression ++ sortColumns)
+        .map(SortOrder(_, Ascending))
+      val actualOrdering = queryExecution.executedPlan.outputOrdering
+      // We can still avoid the sort if the required ordering is [partCol] and the actual ordering
+      // is [partCol, anotherCol].
+      val rdd = if (requiredOrdering == actualOrdering.take(requiredOrdering.length)) {
+        queryExecution.toRdd
+      } else {
+        SortExec(requiredOrdering, global = false, queryExecution.executedPlan).execute()
```

Comment:
Oh, I met this case before IIRC. This complains in Scala 2.10. I guess it should be 

```
SortExec(requiredOrdering, global = false, child = queryExecution.executedPlan).execute()
```

because it seems the complier gets confused the positional/named  arguments. (this is actually invalid syntax in Python).